BUG: Fix rmul dispatch of spmatrix

toslunar opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue

Closes #15210 on error class of `object * spmatrix`.

#### What does this implement/fix?

This PR reverts [this line](https://github.com/scipy/scipy/pull/14822/files#diff-8aed1ba485f8d01bc74679865b6ca8fa04efde8b6fef5a57f587b61eb7a8734cR607) but uses `_mul_dispatch` instead of `*` because `other.transpose().__rmul__` is rarely useful after `other.__mul__` returns NotImplemented. (https://github.com/scipy/scipy/issues/15210#issuecomment-995924719)
This depends on #15232 in order to make sure the fix will not interfere rmul of sparse *array* classes.

#### Additional information
<!--Any additional information you think is important.-->

tylerjereddy commented:
maybe @stefanv or @perimosocordiae can confirm this is needed on top of gh-15232? I suppose this can be rebased now that that is merged to simplify the diff as well.

jarrodmillman commented on scipy/sparse/_arrays.py:
```diff
@@ -46,6 +46,10 @@ def _lil_container(self):
+    def __rmul__(self, *args, **kwargs):
+        return self.multiply(*args, **kwargs)
+
+
```
Too many blank lines (3).

tylerjereddy replied:
ah, that was my fault for merging the original fix without removing the extra line--I thought I could safely ignore the CI complaint because we do have some linting that only applies to new code/diffs, but apparently that was not the case for this one

tylerjereddy replied:
it does indeed make sense to include that fix in this PR if we're going to proceed here

stefanv replied:
I added that fix to this PR.

stefanv commented:
I am not sure the rmul fix proposed here is necessary to get the behavior required. Simply adding rmul as in my merged PR should already have resolved it.

stefanv commented on scipy/sparse/_base.py:
```diff
@@ -604,7 +604,7 @@ def _rmul_dispatch(self, other):
             except AttributeError:
                 tr = np.asarray(other).transpose()
-            return (self.transpose() @ tr).transpose()
+            return self.transpose()._mul_dispatch(tr).transpose()
```
I was curious whether this is to prevent a recursive loop, but I don't think that can happen since this goes to `__mul__`. But I'm fine adding it as a precaution.

toslunar replied:
I propose using `dot` here because I'm not sure if `self.transpose()` must preserve `_is_array` flag especially for user-defined sparse classes.  I'll try to fix #15258, too.
```suggestion
            return self.transpose().dot(tr).transpose()
```

toslunar replied:
Sorry.  `_mul_dispatch` is better because it can return `NotImplemented`.  Now I see `_mul_dispatch` is always "matmul" (or `_mul_scalar`).

stefanv commented:
I rebased the branch and pushed to make the changes proposed easier to read.

I don't know why Github doesn't show a diff between PR branch and master :thinking:

stefanv commented on scipy/sparse/tests/test_base.py:
```diff
@@ -1354,6 +1354,16 @@ def check(dtype):
 
+    # github issue #15210
+    def test_rmul_scalar_type_error(self):
+        def check(dtype):
```
Not sure it's important to test for different dtypes here; important to test for all matrix container types.

tylerjereddy commented:
I removed the extra blank line when I backported the original patch--it looks like there may still be outstanding reviewer comments here, so I'm unlikely to merge it before rc2 happens.

toslunar commented:
I rebased the branch again and squashed some commits.

stefanv commented:
@toslunar Just to orientate myself: could you briefly explain why this change is needed still and what it fixes?

toslunar commented:
### Fixing #15210

The change from `@` to `_mul_dispatch` is to skip the scalar check: for example, `isscalarlike(None)` is False but `isscalarlike(numpy.asarray(None))` is True.
https://github.com/scipy/scipy/blob/4871f3d1c61bdb296ae03e3480f5f584f5c67256/scipy/sparse/_sputils.py#L193-L195


By the way, https://numpy.org/doc/stable/reference/generated/numpy.isscalar.html says that in most cases `np.ndim(x) == 0` should be used instead.  In scipy.sparse

https://github.com/scipy/scipy/blob/4871f3d1c61bdb296ae03e3480f5f584f5c67256/scipy/sparse/_base.py#L539-L542 handles non-numeric cases, but `isscalarlike` might have been defined in a better way.

The branch `if isscalarlike(other)` remains in `_rmul_dispatch` as a fast path.

### Fixing #15258

It is a independent fix now.  On `self.dot(other)`, it's better to use Python operator (`*` or `@`) because it's possible that `other`'s method should be called first.  `@` is sufficient for non-scalar `other`.  For scalar `other`, `self * other` always returns `self._mul_scalar(other)` (unless there's a "scalar" subclass of sparse matrix/array).

toslunar commented:
#15258 is fixed by #16035.

I'll revert my `dot` fixes and merge `main` branch.

perimosocordiae reviewed (approved):
This change looks fine to me, just needs a rebase + possible merge conflict resolution.

toslunar commented:
Rebased the branch and removed the redundant commits:
- c5cb4b1d3 2021-12-21 19:21:53 +0900 BUG: fix sparse dot with scalar
- f4f0e1285 2022-04-28 13:50:49 +0900 REV: "BUG: fix sparse dot with scalar

perimosocordiae commented:
Merging at long last, thanks for your work and patience @toslunar !
